libraries/common/ref.effekt

```diff
@@ -9,26 +9,69 @@ extern js """
   }
 """
 
+extern llvm """
+define void @c_ref_erase_field(ptr %0) {
+  %field_pointer = getelementptr inbounds ptr, ptr %0, i64 1
+  %tag = load i64, ptr %0, align 8
+  %field = load ptr, ptr %field_pointer, align 8
+  tail call void @erasePositive(i64 %tag, ptr %field)
+  ret void
+}
+"""
+
 /// Global, mutable references
 extern type Ref[T]
 
+/** We represent references like positive types in LLVM
+ *  The tag is 0 and the obj points to memory with the following layout:
+ *
+ *   +--[ Header ]--------------+------------+
+ *   | ReferenceCount  | Eraser | Field      |
+ *   +--------------------------+------------+
+ */
+
 /// Allocates a new reference, keeping its value _undefined_.
 /// Prefer using `ref` constructor instead to ensure that the value is defined.
 extern global def allocate[T](): Ref[T] =
   js "{ value: undefined }"
   chez "(box #f)"
   llvm """
-    %z = call %Pos @c_ref_fresh(%Pos zeroinitializer)
-    ret %Pos %z
+    ; sizeof Header + sizeof Pos = 32
+    %ref = tail call noalias ptr @malloc(i64 noundef 32)
+    %refEraser = getelementptr ptr, ptr %ref, i64 1
+    %fieldTag = getelementptr ptr, ptr %ref, i64 2
+    %fieldData_pointer = getelementptr ptr, ptr %ref, i64 3
+
+    store i64 0, ptr %ref, align 8
+    store ptr @c_ref_erase_field, ptr %refEraser, align 8
+    store i64 0, ptr %fieldTag, align 8
+    store ptr null, ptr %fieldData_pointer, align 8 
+
+    %refWrap = insertvalue %Pos { i64 0, ptr poison }, ptr %ref, 1
+    ret %Pos %refWrap
   """
 
 /// Creates a new reference with the initial value `init`.
 extern global def ref[T](init: T): Ref[T] =
   js "{ value: ${init} }"
   chez "(box ${init})"
   llvm """
-    %z = call %Pos @c_ref_fresh(%Pos ${init})
-    ret %Pos %z
+    %initTag = extractvalue %Pos ${init}, 0
+    %initObject_pointer = extractvalue %Pos ${init}, 1
+    ; sizeof Header + sizeof Pos = 32
+    %ref = tail call noalias ptr @malloc(i64 noundef 32)
+    %refEraser = getelementptr ptr, ptr %ref, i64 1
+    %fieldTag = getelementptr ptr, ptr %ref, i64 2
+    %fieldData_pointer = getelementptr ptr, ptr %ref, i64 3
```

Comment:
Here too I believe these could be 1 `getelementptr` and later 1 `store` without extracting the fields of `init`.